basil/HL/bronkhorst_elflow.py

```diff
@@ -16,160 +17,100 @@
 
 class Bronkhorst_ELFLOW(HardwareLayer):
     ''' Bronkhorst ELFLOW
+    Manual can be found here:
+    https://www.bronkhorst.com/getmedia/77a1438f-e547-4a79-95ad-53e81fd38a97/917027-Manual-RS232-interface.pdf
     '''
 
+    CMDS = {
+        'get_measure_flow': ':06800401210120',
+        'get_capacity': ':068004014D014D',
+        'get_control_mode': ':06800401040104',
+        'set_control_mode': ':0580010104',
+        'set_setpoint': ':0680010121',
+        'get_setpoint': ':06800401210121',
+        'get_valve': ':06800472417241',
+    }
+
     def __init__(self, intf, conf):
         self.debug = 0
         self.node = "80"
         super(Bronkhorst_ELFLOW, self).__init__(intf, conf)
+        self.pre_time = time.time()
 
     def init(self):
         super(Bronkhorst_ELFLOW, self).init()
 
     def write(self, cmd):
-        cmd_s = ""
-        for c in cmd:
-            cmd_s = cmd_s + "%02X" % c
-        cmd_s = ":%02X%s%s" % (len(cmd) + 1, self.node, cmd_s)
-        if self.debug != 0:
-            logger.debug("ELFLOW.write() %s" % str(cmd_s))
-        self._intf.write(cmd_s)
+        if time.time() - self.pre_time < 1.0:
+            time.sleep(1.0)
+        self._intf.write(str(cmd))
+        self.pre_time = time.time()
 
     def read(self):
-        ret_s = self._intf.read()
-        if self.debug != 0:
-            logger.debug("ELFLOW.read() %s" % str(ret_s))
-        if len(ret_s) < 5 or ret_s[0] != ":" or ret_s[3:5] != self.node:
-            logger.debug("ELFLOW.read() format error ret=%s" % str(ret_s))
-            return []
-        ret_len = int(ret_s[1:3])
-        if ret_len * 2 != len(ret_s[3:-2]):
-            logger.debug("ELFLOW.read() data lenth error ret=%s" % str(ret_s))
-            return []
-        ret = []
-        for i in range(ret_len - 1):
-            ret.append(int(ret_s[5 + 2 * i:5 + 2 * (i + 1)], 16))
-        return ret
+        ret = self._intf.read()
+        if len(ret) < 2 or ret[-2:] != "\r\n":
+            logger.warning("read() termination error")
+        return ret.strip()
+
+    def get_valve_output(self):
+        self._intf.write(self.CMDS['get_valve'])
+        ret = int(self.read()[11:], 16)
+        return ret * 61.7 / 10345949  # converts int in percentage
 
     def set_setpoint(self, value):
-        cmd = [1, 1, 0x21, (value >> 8) & 0xFF, value & 0xFF]
-        self.write(cmd)
+        """value range from 0 - 32000
```

Comment:
Is the range exactly 32000 or 32767 aka 2^15? The command seems to allow 2^16 actually? Iirc, there is a large range that does not influence the effective the setting, is it this one? Not sure if we should print a warning or info or ignore it, since the user should not care? @YannickDieter

Reply:
I heard @SilasM2001, you know more about that?

Reply:
I worked with the new version in the flowmeter branch. I can not commend on the previous code. The manual says setpoint has a range of 0 to 32000, with 32000 being 100%. In my experience the range works fine. The module testing setup of Matthias uses a large range (1000-25000 or so, if i recall correctly) and i did not observe a large range that does not influence the effective setting.

Reply:
I just wanted to say again: The manual says setpoint has a range from 0 to 32000, 32000 being 100%. I dont know why the old command allowed values up to 32767.